pkg/cloudproduct/gke/gke.go

```diff
@@ -82,3 +88,74 @@ func (*gkeAutopilot) SyncPodPortsToGameServer(gs *agonesv1.GameServer, pod *core
 	}
 	return nil
 }
+
+func (*gkeAutopilot) PortAllocatorFactory(minPort, maxPort int32,
+	_ informers.SharedInformerFactory,
+	_ externalversions.SharedInformerFactory) portallocator.Interface {
+	return &autopilotPortAllocator{minPort: minPort, maxPort: maxPort}
+}
+
+func (*gkeAutopilot) ValidateGameServer(gs *agonesv1.GameServer) []metav1.StatusCause {
+	var causes []metav1.StatusCause
+	for _, p := range gs.Spec.Ports {
+		if p.PortPolicy != agonesv1.Dynamic {
+			causes = append(causes, metav1.StatusCause{
+				Type:    metav1.CauseTypeFieldValueInvalid,
+				Field:   fmt.Sprintf("%s.portPolicy", p.Name),
+				Message: errPortPolicyMustBeDynamic,
+			})
+		}
+	}
+	return causes
+}
+
+type autopilotPortAllocator struct {
+	minPort int32
+	maxPort int32
+}
+
+func (*autopilotPortAllocator) Run(_ context.Context) error        { return nil }
+func (*autopilotPortAllocator) DeAllocate(gs *agonesv1.GameServer) {}
+
+func (apa *autopilotPortAllocator) Allocate(gs *agonesv1.GameServer) *agonesv1.GameServer {
+	if len(gs.Spec.Ports) == 0 {
+		return gs // Nothing to do.
+	}
+
+	var ports []agonesv1.GameServerPort
+	for i, p := range gs.Spec.Ports {
+		if p.PortPolicy != agonesv1.Dynamic {
+			logger.Errorf("GameServer %s port %v has PortPolicy %q - this should have been rejected by webhooks, refusing to assign ports",
+				gs.Name, p, p.PortPolicy)
+			return gs
+		}
+		p.HostPort = int32(i + 1) // Autopilot expects _some_ host port - use a value unique to this GameServer Port.
```

Comment:
A thought I just had - not for this PR, but eventually we'll need a way to turn off some e2e tests for Autopilot (and maybe even add some Autopilot specific ones?)

Just something to think about.

Reply:
Yeah, I filed an internal bug to get e2es working on AP - I'll file an issue here soon, too.